src/tests/testCacheManager.cc

```diff
@@ -66,3 +75,86 @@ testCacheManager::testRegister()
     CPPUNIT_ASSERT_EQUAL(1,(int)sentry->flags);
 }
 
+void
+testCacheManager::testParseUrl()
+{
+    auto *mgr = static_cast<CacheManagerInternals *>(CacheManager::GetInstance());
+    CPPUNIT_ASSERT(mgr != nullptr);
+
+    std::vector<AnyP::ProtocolType> validSchemes = {
+        AnyP::PROTO_CACHE_OBJECT,
+        AnyP::PROTO_HTTP,
+        AnyP::PROTO_HTTPS,
+        AnyP::PROTO_FTP
+    };
+
+    AnyP::Uri mgrUrl;
+    mgrUrl.host("localhost");
+    mgrUrl.port(3128);
+
+    const std::vector<const char *> validPathActions = {
+        // "", // technically valid, but not supported
+        "/",
+        "/menu",
+        // "/squid-internal-mgr", // technically valid, but not supported
+        "/squid-internal-mgr/",
+        "/squid-internal-mgr/menu"
+    };
+
+    const std::vector<const char *> validParams = {
+        "",
+        "?",
+        "?&",
+        "?&&&&&&&&&&&&",
+        "?foo=bar",
+        "?0123456789=bar",
+        "?foo=bar&",
+        "?foo=bar&&&&",
+        "?&foo=bar",
+        "?&&&&foo=bar",
+        "?&foo=bar&",
+        "?&&&&foo=bar&&&&",
+        "?foo=?_weird?~`:[]stuff&bar=okay&&&&&&"
+    };
+
+    const std::vector<const char *> validFragments = {
+        "",
+        "#frag"
+    };
+
+    unsigned caseNumber = 0;
+    unsigned success = 0;
+    for (const auto &scheme : validSchemes) {
+        mgrUrl.setScheme(scheme);
+
+        for (const auto *action : validPathActions) {
+
+            // all schemes except cache_object require "/squid-internal-mgr" path prefix
+            if (scheme != AnyP::PROTO_CACHE_OBJECT && strlen(action) <= 19)
+                continue;
+
+            for (const auto *param : validParams) {
+
+                for (const auto *frag : validFragments) {
+                    try {
+                        ++caseNumber;
+                        std::cerr << caseNumber << std::endl;
+
+                        SBuf bits;
+                        bits.append(action);
+                        bits.append(param);
+                        bits.append(frag);
+                        mgrUrl.path(bits);
+
+                        (void)mgr->ParseUrl(mgrUrl);
+                        ++success;
+                    } catch (...) {
+                        std::cerr << std::endl << "FAIL: " << mgrUrl << std::endl
+                                  << CurrentException << std::endl;
```

Comment:
Please avoid unqualified/unmarked/unprefixed exceptions. The resulting output can be confusing, and this source code may be very difficult to find when things go seriously wrong:

```suggestion
                                  << "    error: " << CurrentException << std::endl;
```

Reply:
CurrentException prints its own standard prefix. The qualification marking is the FAIL line immediately above and the name of the test it is running. The output to the test-suite log looked good to me, will post an example later so you can see.

Reply:
> CurrentException prints its own standard prefix

It does not. And it should not. CurrentException does not know its reporting context so it should not supply any prefix. It is the caller responsibility, and that responsibility is neglected in this code.

> The qualification marking is the FAIL line immediately above

Yes, but the "FAIL" line does not help in filtered results (such as those often pasted in support emails), especially when other test cases produce misleading FAIL lines, and especially since this output does not follow the emerging multiline standard introduced by Debug::Extra -- there will be no visual clue that something important was filtered out. Visually, the reported part of the log will just contain a seemingly random message on its own line. When things go wrong, that line may cause additional headaches. The minimal fix is trivial and is provided. We should not spend time discussing why it should be avoided!

Reply:
Oh you are right. I was thinking of the "exception location:" prefix.

Failure display updated in 351e7ba56. On failures it will now produce something like:
```
...
FAIL: cache_object://localhost/?foo= bar
    error: missing or malformed parameter value
    exception location: src/mgr/QueryParams.cc(129) Parse
F
testCacheManager.cc:157:Assertion
Test name: testCacheManager::testParseUrl
equality assertion failed
- Expected: 21
- Actual  : 20
```